Update google.golang.org/grpc to 1.68.0

espadolini opened the pull request:
Starting from grpc-go 1.67, TLS transport credentials (on both server and client side) enforce the use of ALPN - however, the initial concern that led us to not upgrade to 1.67 immediately, i.e. the fact that we often use custom ALPN values, is not a problem here: as long as client and server negotiate _some_ ALPN, the connection will still succeed even with the transport credentials implementation in grpc-go; moreover, starting from 1.68.0, grpc-go will actually correctly apply its modifications to the `tls.Config`s in use even when making use of `GetConfigForClient` (even though all of our uses of configs with `GetConfigForClient` should already be setting `NextProtos` correctly).

The `go` version in the API module is bumped from 1.22.0 to 1.22.7 to accomodate grpc-go's new minimum version.

espadolini committed:
Update google.golang.org/grpc to 1.68.0

codingllama reviewed (commented):
Thanks, Edoardo.

No testing changes needed? I vaguely remember something breaking.

tigrato reviewed (approved):
Thanks!

espadolini committed:
Fix TestTransportCredentials_ServerHandshake

espadolini committed:
Fix TestStart

The credentials.NewTLS call will add the h2 NextProto if missing, but in
some tests we create a tls.Config and use it to connect to the test
server directly, so it's cleaner to just create the tls.Config with the
correct NextProto in the first place.

codingllama commented on lib/auth/transport_credentials_test.go:
```diff
@@ -286,7 +287,13 @@ func TestTransportCredentials_ServerHandshake(t *testing.T) {
+			// this would be done by the grpc TransportCredential in the client
+			clientTLSConf := test.clientTLSConf
+			if !slices.Contains(clientTLSConf.NextProtos, "h2") {
+				clientTLSConf = clientTLSConf.Clone()
```
Do we need the Clone call?

espadolini replied:
Seems rude not to?

codingllama commented on lib/auth/transport_credentials_test.go:
```diff
@@ -286,7 +287,13 @@ func TestTransportCredentials_ServerHandshake(t *testing.T) {
 			t.Cleanup(func() { require.NoError(t, conn.Close()) })
 
-			clientConn := tls.Client(conn, test.clientTLSConf)
+			// this would be done by the grpc TransportCredential in the client
```
Add some context regarding grpc-go v1.68.0 and the need for the "h2" NextProto?

codingllama commented on lib/teleterm/grpccredentials.go:
```diff
@@ -118,6 +118,7 @@ func createClientTLSConfig(clientKeyPair tls.Certificate, serverCertPath string)
 
 	return &tls.Config{
 		Certificates: []tls.Certificate{clientKeyPair},
+		NextProtos:   []string{"h2"},
```
Is there a const we could use for "h2"?

Could you add some context on why the NextProtos is necessary here too?

espadolini replied:
There's a const in x/net/http2 but it's hardly ever used, even in library code.

I've actually moved this to the test code as well, since we hardly ever manually add "h2" ourselves in the rest of the codebase.

espadolini committed:
Tweaks to address review comments
